homeassistant/components/qnap/__init__.py

```diff
@@ -1 +1,129 @@
+from homeassistant.helpers import config_per_platform, device_registry as dr
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator
+
+from .const import COMPONENTS, DEFAULT_NAME, DEFAULT_PORT, DEFAULT_TIMEOUT, DOMAIN
+
+UPDATE_INTERVAL = timedelta(minutes=1)
+
+_LOGGER = logging.getLogger(__name__)
+
+
+async def async_setup(hass, config):
+    """Set up the qnap environment."""
+    hass.data.setdefault(DOMAIN, {})
+
+    # Import configuration from sensor platform
+    config_platform = config_per_platform(config, "sensor")
+    for p_type, p_config in config_platform:
+        if p_type != DOMAIN:
+            continue
+
+        hass.async_create_task(
+            hass.config_entries.flow.async_init(
+                DOMAIN,
+                context={"source": config_entries.SOURCE_IMPORT},
+                data=p_config,
+            )
+        )
+
+    return True
+
+
+async def async_setup_entry(hass, config_entry):
+    """Set the config entry up."""
+    host = config_entry.data[CONF_HOST]
+    protocol = "https" if config_entry.options.get(CONF_SSL) else "http"
+    api = QNAPStats(
+        host=f"{protocol}://{host}",
+        port=config_entry.data.get(CONF_PORT, DEFAULT_PORT),
+        username=config_entry.data[CONF_USERNAME],
+        password=config_entry.data[CONF_PASSWORD],
+        verify_ssl=config_entry.options.get(CONF_VERIFY_SSL),
+        timeout=DEFAULT_TIMEOUT,
+    )
+    try:
+        system_info = await hass.async_add_executor_job(api.get_system_stats)
+    except ConnectTimeout as error:
+        raise PlatformNotReady from error
+
+    device_registry = await dr.async_get_registry(hass)
+    device_registry.async_get_or_create(
+        config_entry_id=config_entry.entry_id,
+        identifiers={(DOMAIN, config_entry.unique_id)},
+        manufacturer=DEFAULT_NAME,
+        name=system_info.get("system", {}).get("name", host),
+        model=system_info.get("system", {}).get("model"),
+        sw_version=system_info.get("firmware", {}).get("version"),
+    )
+
+    async def async_update_data():
+        datas = {}
+        datas["system_stats"] = await hass.async_add_executor_job(api.get_system_stats)
+        datas["system_health"] = await hass.async_add_executor_job(
+            api.get_system_health
+        )
+        datas["smart_drive_health"] = await hass.async_add_executor_job(
+            api.get_smart_disk_health
+        )
+        datas["volumes"] = await hass.async_add_executor_job(api.get_volumes)
+        datas["bandwidth"] = await hass.async_add_executor_job(api.get_bandwidth)
+        return datas
+
+    coordinator = DataUpdateCoordinator(
+        hass,
+        _LOGGER,
+        name="sensor",
+        update_method=async_update_data,
+        update_interval=UPDATE_INTERVAL,
+    )
+
+    # Fetch initial data so we have data when entities subscribe
+    await coordinator.async_refresh()
+
+    hass.data[DOMAIN] = coordinator
```

Comment:
What if the user configures two instances of the integration? The second instance `coordinator` will replace the first instance `coordinator`. You can use `hass.data[DOMAIN][config_entry.entry_id] = coordinator` to avoid this.